elasticsearch_helpers/ElasticsearchClient.go

```diff
@@ -0,0 +1,85 @@
+package elasticsearch_helpers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"github.com/elastic/go-elasticsearch/v8"
+	"log"
+)
+
+type ElasticsearchClient struct {
+	client *elasticsearch.Client
+}
+
+func GetElasticsearchClient() ElasticsearchClient {
+	cfg := elasticsearch.Config{
+		Addresses: []string{
+			"http://localhost:9200",
+		},
+	}
+	es, err := elasticsearch.NewClient(cfg)
+	if err != nil {
+		log.Fatalf("Error getting client: %s", err)
+	}
+	return ElasticsearchClient{
+		client: es,
+	}
+}
+
+func (esclient *ElasticsearchClient) SearchDocument(term string) []byte {
+	var buf bytes.Buffer
+	query := map[string]interface{}{
+		"query": map[string]interface{}{
+			"wildcard": map[string]interface{}{
+				"name": "*" + term + "*",
+			},
+		},
+	}
+	if err := json.NewEncoder(&buf).Encode(query); err != nil {
+		log.Fatalf("Error encoding query: %s", err)
+	}
+
+	res, err := esclient.client.Search(
+		esclient.client.Search.WithContext(context.Background()),
+		esclient.client.Search.WithIndex("product"),
+		esclient.client.Search.WithBody(&buf),
+		esclient.client.Search.WithTrackTotalHits(true),
+		esclient.client.Search.WithPretty(),
+	)
+	if err != nil {
+		log.Fatalf("Error getting response: %s", err)
+	}
+	defer res.Body.Close()
+
+	if res.IsError() {
+		var e map[string]interface{}
+		if err := json.NewDecoder(res.Body).Decode(&e); err != nil {
+			log.Fatalf("Error parsing the response body: %s", err)
+		} else {
+			// Print the response status and error information.
+			log.Fatalf("[%s] %s: %s",
+				res.Status(),
+				e["error"].(map[string]interface{})["type"],
+				e["error"].(map[string]interface{})["reason"],
+			)
+		}
+	}
+
+	var r map[string]interface{}
+
+	if err := json.NewDecoder(res.Body).Decode(&r); err != nil {
+		log.Fatalf("Error parsing the response body: %s", err)
+	}
+
+	var final []map[string]interface{}
+
+	// Print the ID and document source for each hit.
+	for _, hit := range r["hits"].(map[string]interface{})["hits"].([]interface{}) {
```

Comment:
Check the latest version, I think we've covered this what you've said.

Reply:
It still looks the same ( maybe forgot to push that part :) ) Still, this looks very confusing. And whenever you're working with a map and access data like this, you risk getting an index out of range exception. Go provides a nice mechanism to check if key is present in the map. Have you maybe considered creating a helper struct for unmarshalling this payload?

Reply:
I've added an assertion check in the form of "`result, ok := r["hits"].(map[string]interface{})["hits"].([]interface{})`". Then, the extraction of data will only proceed if "ok" is true.

No, we haven't thought of that, simply because that payload is full of unnecessary information for our use case, therefore we skipped it and extracted only what we needed.